### ISSUE TITLE
[Objective-C] Assertion "Narrow integer argument must have a valid extension type." failed.

### ISSUE THREAD

**Issue by JonPsson1** [llvm/llvm-project#109654]
With the recent strengthening of the handling of extensions of narrow integer arguments (commit 1412022), it is now possible to detect cases of missing argument extensions, although there are still some issues that keeps this disabled by default. One of these test failures is 

clang/test/CodeGenObjC/blocks-ivar-debug.m -v

What this means is that there is a narrow (<=32 bit) integer argument that is neither sign or zero extended, and is also not marked as noext, which should be done for "struct-in-reg". So this may or may not be a "bug" (missing sign/zero extension), or it could be an argument that should be marked as noext. See https://llvm.org/docs/LangRef.html#parameter-attributes.

This can be reproduced with:

clang -cc1 -internal-isystem ~/llvm-project/build/lib/clang/20/include ~/llvm-project/clang/test/CodeGenObjC/blocks-ivar-debug.m -fblocks -S -o /dev/null -triple s390x--linux-gnu -mllvm -argext-abi-check

llvm/lib/Target/SystemZ/SystemZISelLowering.cpp:9861: void llvm::SystemZTargetLowering::verifyNarrowIntegerArgs(const llvm::SmallVectorImpl<llvm::ISD::OutputArg>&, bool) const: Assertion `(VT != MVT::i32 || (Flags.isSExt() || Flags.isZExt() || Flags.isNoExt())) && "Narrow integer argument must have a valid extension type."' failed.

CC:ing some people that seem to be familiar with ObjC: @kazutakahirata @danix800 , thanks.

**Comment by JonPsson1** [comment]
@nikic

**Comment by nikic** [comment]
@JonPsson1 It would be good if the error also printed the function name and signature. That would probably make the cause a lot more obvious.

**Comment by JonPsson1** [comment]
With https://github.com/llvm/llvm-project/pull/109699, I get:
```

ERROR: Missing extension attribute of passed value in call to function:
Callee:  void @_Block_object_assign(ptr, ptr, i32)
Caller:  void @__copy_helper_block_8_32o(ptr, ptr)

UNREACHABLE executed
```

**Comment by JonPsson1** [comment]
@nikic @kazutakahirata @danix800

**Comment by JonPsson1** [comment]
The improvement to dump the function signature for cases of missing extension attributes has now been committed.

(f9fbfc5)

**Comment by JonPsson1** [comment]
ping...

**Comment by efriedma-quic** [comment]
I think _Block_object_assign is supposed to take a signed integer.

**Comment by JonPsson1** [comment]
Patch proposed: #111740

**Comment by JonPsson1** [comment]
Fixed by #113506
